### DISCUSSION TITLE
How to pass pip args to spacy.cli.download?

### DISCUSSION THREAD

**Discussion by loretoparisi** [explosion/spaCy#8225] [Help: Coding & Implementations] [answered]
I want to  programmatically download models using `spacy.cli`. According to the doc, it is possibile to pass a `--data-path`
to the command line `download` command, hence it should be possibile to pass it to the `spacy.cli.download`:

```python
pip_args = ???
cache_dir=os.getenv("cache_dir", "../../models")
try:
    nlp = spacy.load("en_core_web_sm")
except OSError:
    from spacy.cli import download
    download("en_core_web_sm", pip_args)
    nlp = spacy.load('en')
```

using `download(model: str, direct: bool = False, sdist: bool = False, *pip_args) -&gt; None:<br class="Apple-interchange-newline">`

so how to pass `data-path` as `pip_args` input?

**Comment by polm** [comment]
I'm pretty sure the `download` command doesn't support a `--data-path` argument. That argument is used some other commands, like `evaluate`, but not download. In practice actually running the command that way gives an error. Where did you see that in the docs? 

Note that in this case `pip_args` is just a list of strings that are used as arguments to `pip`, which also doesn't take a `--data-path` arg. 

The path pip saves the model in is just wherever it's installing packages, so you can control that by putting a virtualenv in the right place, for example.

**Reply by loretoparisi** [reply]
Thanks, I did not find a better way than the following

```python
cache_dir=os.getenv("cache_dir", "../../models")
model_path="en_core_web_sm"
try:
    nlp = spacy.load(os.path.join(cache_dir,model_path))
except OSError:
    spacy.cli.download(model_path)
    nlp = spacy.load(model_path)
    nlp.to_disk(os.path.join(cache_dir,model_path))
```

**Reply by polm** [reply]
That is definitely one way to make sure the models are in a specific location. Note that it will result in two copies of the model on disk, one in wherever pip installing packages and one in your `cache_dir`. That may or may not be a problem for you.

**Comment by ines** [comment, marked as answer]
The `spacy download` command will pass all other arguments (that are not pre-defined) forward to `pip install`. This means you can do something like `spacy download en_core_web_sm --no-cache-dir --no-deps`, or add arguments to pass in an SSL key file. Here are the available options to `pip install`: https://pip.pypa.io/en/stable/cli/pip_install/#options I don't think it supports a `--data-path`, though?

In the Python function, this is implemented as variable positional arguments (`*pip_args`). So anything you pass into that function aside from the defined arguments gets passed to `pip`. So you can do:

```python
download("en_core_web_sm", "--no-cache-dir", "--no-deps")
```
